File: C/05_Extras/01_non-blocking_ring_pass.c
#include <stdio.h>
#include <mpi.h>


int main(int argc, char **argv){
  int rank, nproc, rank_right, rank_left, rank_recv;
  MPI_Request requests[2];
  MPI_Init(&argc, &argv);
  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
  MPI_Comm_size(MPI_COMM_WORLD, &nproc);

  rank_left = rank -1;
  /*Ranks run from 0 to nproc-1, so wrap the ends around to make a loop*/
  if(rank_left == -1) rank_left = nproc-1;
  rank_right = rank + 1;
  if(rank_right == nproc) rank_right = 0;

  MPI_Isend(&rank, 1, MPI_INT, rank_right, 100, MPI_COMM_WORLD, &requests[0]);
  MPI_Irecv(&rank_recv, 1, MPI_INT, rank_left, 100, MPI_COMM_WORLD, 
      &requests[1]);

  MPI_Waitall(2, requests, MPI_STATUSES_IGNORE);

  printf("Rank %i has received value %i from rank %i\n", rank, rank_recv,
      rank_left);

  MPI_Finalize();
}

File: Fortran/05_extras/01_non-blocking_ring_pass.f90

PROGRAM main

  USE MPI
  IMPLICIT NONE
  INTEGER :: rank, nproc, rank_right, rank_left, rank_recv, errcode
  INTEGER, DIMENSION(2) :: requests

  CALL MPI_INIT(errcode)
  CALL MPI_COMM_RANK(MPI_COMM_WORLD, rank, errcode)
  CALL MPI_COMM_SIZE(MPI_COMM_WORLD, nproc, errcode)

  rank_left = rank -1
  !Ranks run from 0 to nproc-1, so wrap the ends around to make a loop
  IF(rank_left == -1) rank_left = nproc-1
  rank_right = rank + 1
  IF(rank_right == nproc) rank_right = 0

  CALL MPI_Isend(rank, 1, MPI_INTEGER, rank_right, 100, MPI_COMM_WORLD, &
      requests(1), errcode)
  CALL MPI_Irecv(rank_recv, 1, MPI_INTEGER, rank_left, 100, MPI_COMM_WORLD, &
      requests(2), errcode)

  !We don't actually care about sending being finished here, but may as well
  !wait for both
  CALL MPI_Waitall(2, requests, MPI_STATUSES_IGNORE, errcode)

  PRINT ('("Rank ",I3," has received value ", I3, " from rank ", I3)'), &
      rank, rank_recv, rank_left

  CALL MPI_FINALIZE(errcode)
END PROGRAM main
